File: src/services/clickup.js
const axios = require('axios');
const FormData = require('form-data');
const logger = require('../config/logger');

class ClickUpService {
  constructor() {
    this.token = process.env.CLICKUP_TOKEN;
    this.listId = process.env.CLICKUP_LIST_ID;
    this.client = axios.create({
      baseURL: 'https://api.clickup.com/api/v2',
      headers: {
        'Authorization': this.token
      }
    });
  }

  async uploadImage(taskId, imageBuffer, filename) {
    try {
      const formData = new FormData();
      formData.append('attachment', imageBuffer, {
        filename: filename,
        contentType: 'image/jpeg'
      });

      const uploadResponse = await this.client.post(
        `/task/${taskId}/attachment`,
        formData,
        {
          headers: {
            ...formData.getHeaders()
          }
        }
      );

      return uploadResponse.data;
    } catch (error) {
      logger.error('ClickUp image upload failed:', {
        error: error.response?.data || error.message,
        taskId,
        filename
      });
      throw new Error('Failed to upload image to ClickUp');
    }
  }

  async createTask(id, imageBuffer, filename) {
    try {
      // First create the task
      const taskResponse = await this.client.post(`/list/${this.listId}/task`, {
        name: `Image Processing Task - ${id}`,
        description: `Image Reference ID: ${id}`,
        priority: 3,
        status: 'to do'
      });

      const taskId = taskResponse.data.id;
      logger.info(`Created ClickUp task: ${taskId}`);

      // Then upload and attach the image
      await this.uploadImage(taskId, imageBuffer, filename);
      logger.info(`Uploaded image to task: ${taskId}`);
      
      return taskResponse.data;
    } catch (error) {
      logger.error('ClickUp task creation failed:', {
        error: error.response?.data || error.message,
        id
      });
      throw new Error(`Failed to create ClickUp task: ${error.response?.data?.err || error.message}`);
    }
  }
}

module.exports = new ClickUpService();

File: utils/clickup.py
import os

import requests

class ClickUpClient:
    def __init__(self):
        self.token = os.getenv('CLICKUP_TOKEN')
        self.list_id = os.getenv('CLICKUP_LIST_ID')
        self.team_id = os.getenv('CLICKUP_TEAM_ID')
        self.base_url = "https://api.clickup.com/api/v2"

    def create_task(self, attachment, id, description:str=None):
        url = f"{self.base_url}/list/{self.list_id}/task"

        headers = {
            "Content-Type": "application/json",
            "Authorization": self.token
        }
        query = {
            "custom_task_ids": "true",
            "team_id": self.team_id
        }
        payload = {
            "name": f"Ticket_{id}",
            "description": f"ID: {id}\n\nDescription: {description}",
            "assignees": [
                #89657945
            ],
        }

        response = requests.post(url, json=payload, headers=headers, params=query)
        data = response.json()

        if attachment and attachment != "":
            self.add_attachment(data['id'], attachment, True)
        
        return data

    def add_attachment(self, task_id, attachment, delete_attachment=False):
        url = f"{self.base_url}/task/{task_id}/attachment"
        headers = {
            "Authorization": self.token
        }
        file = {
            "attachment": (attachment, open(attachment, 'rb'))
        }

        response = requests.post(url, files=file, headers=headers)

        if delete_attachment:
            os.remove(attachment)
        
        return response.json()